File: 08.string-to-integer-atoi/string-to-integer-atoi.go
package main

import (
	"fmt"
	"math"
	"strings"
	"unicode"
)

func atoi(s string) int {
	i := 0
	for _, c := range s {
		if c == ' ' {
			i++
		}
	}

	s = s[i:]

	sign := 1
	if strings.HasPrefix(s, "-") {
		sign = -1
		s = s[1:]
	} else if strings.HasPrefix(s, "+") {
		sign = 1
		s = s[1:]
	}

	result := 0

	for _, digit := range s {
		if unicode.IsDigit(digit) {
			result = result*10 + (int(digit) - int('0'))
		} else {
			break
		}

		if result*sign >= math.MaxInt32 {
			return math.MaxInt32
		} else if result*sign <= math.MinInt32 {
			return math.MinInt32
		}
	}

	return result * sign
}

func solveProblem(s string) {
	fmt.Printf("string: %s, number: %d\n", s, atoi(s))
}

func main() {
	solveProblem("")
	solveProblem("42")
	solveProblem("   -42")
	solveProblem("4193 with words")
	solveProblem("words and 987")
	solveProblem("-91283472332")
}

File: 08.string-to-integer-atoi/string-to-integer-atoi.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

MIN_INT32 = -0x80000000
MAX_INT32 = 0x7fffffff


def atoi(s: str) -> int:
    i = 0
    for _, c in enumerate(s):
        if c == " ":
            i += 1
        else:
            break
    s = s[i:]
    sign = 1
    if s.startswith("-"):
        sign = -1
        s = s[1:]
    elif s.startswith("+"):
        sign = 1
        s = s[1:]

    result = 0
    for _, c in enumerate(s):
        if c.isdigit():
            result = result * 10 + (int(c))
        else:
            break
        if result * sign >= MAX_INT32:
            return MAX_INT32
        if result * sign <= MIN_INT32:
            return MIN_INT32

    return result * sign


def solve_problem(s: str) -> None:
    print(f"string: {s}, integer: {atoi(s)}")


if __name__ == '__main__':
    solve_problem("")
    solve_problem("42")
    solve_problem("   -42")
    solve_problem("4193 with words")
    solve_problem("words and 987")
    solve_problem("-91283472332")
